plugins/inputs/hugepages/hugepages.go

```diff
@@ -0,0 +1,292 @@
+//go:build linux
+// +build linux
+
+package hugepages
+
+import (
+	"bytes"
+	"fmt"
+	"io/ioutil"
+	"path/filepath"
+	"strconv"
+	"strings"
+
+	"github.com/influxdata/telegraf"
+	"github.com/influxdata/telegraf/plugins/inputs"
+)
+
+const (
+	// path to root huge page control directory
+	rootHugepagePath = "/sys/kernel/mm/hugepages"
+	// path where per NUMA node statistics are kept
+	numaNodePath = "/sys/devices/system/node"
+	// path to the meminfo file
+	meminfoPath = "/proc/meminfo"
+
+	rootHugepages    = "root"
+	perNodeHugepages = "per_node"
+	meminfoHugepages = "meminfo"
+
+	hugepagesSampleConfig = `
+  ## Supported huge page types:
+  ##   - "root" - based on root huge page control directory: /sys/kernel/mm/hugepages
+  ##   - "per_node" - based on per NUMA node directories: /sys/devices/system/node/node[0-9]*/hugepages
+  ##   - "meminfo" - based on /proc/meminfo file
+  # types = ["root", "per_node"]
+`
+)
+
+var (
+	newlineByte = []byte("\n")
+	colonByte   = []byte(":")
+
+	hugepagesMetricsRoot = map[string]string{
+		"free_hugepages":          "free",
+		"nr_hugepages":            "total",
+		"nr_hugepages_mempolicy":  "mempolicy",
+		"nr_overcommit_hugepages": "overcommit",
+		"resv_hugepages":          "reserved",
+		"surplus_hugepages":       "surplus",
+	}
+
+	hugepagesMetricsPerNUMANode = map[string]string{
+		"free_hugepages":    "free",
+		"nr_hugepages":      "total",
+		"surplus_hugepages": "surplus",
+	}
+
+	hugepagesMetricsFromMeminfo = map[string]string{
+		"HugePages_Total": "total_of_default_size",
+		"HugePages_Free":  "free_of_default_size",
+		"HugePages_Rsvd":  "reserved_of_default_size",
+		"HugePages_Surp":  "surplus_of_default_size",
+		"Hugepagesize":    "default_size_kb",
+		"Hugetlb":         "total_consumed_by_all_sizes_kb",
+		"AnonHugePages":   "anonymous_kb",
+		"ShmemHugePages":  "shared_memory_kb",
+		"FileHugePages":   "file_kb",
+	}
+)
+
+type Hugepages struct {
+	Types []string `toml:"types"`
+
+	gatherRoot    bool
+	gatherPerNode bool
+	gatherMeminfo bool
+
+	rootHugepagePath string
+	numaNodePath     string
+	meminfoPath      string
+}
+
+func (h *Hugepages) Description() string {
+	return "Gathers huge pages measurements."
+}
+
+func (h *Hugepages) SampleConfig() string {
+	return hugepagesSampleConfig
+}
+
+func (h *Hugepages) Init() error {
+	err := h.parseHugepagesConfig()
+	if err != nil {
+		return err
+	}
+
+	h.rootHugepagePath = rootHugepagePath
+	h.numaNodePath = numaNodePath
+	h.meminfoPath = meminfoPath
+
+	return nil
+}
+
+func (h *Hugepages) Gather(acc telegraf.Accumulator) error {
+	var err error
+
+	if h.gatherRoot {
+		err = h.gatherRootStats(acc)
+		if err != nil {
+			return fmt.Errorf("gathering root stats failed: %v", err)
+		}
+	}
```

Comment:
There is no need to declare `err` here as it always lives in the `if` scope(s)...
```suggestion
	if h.gatherRoot {
		if err := h.gatherRootStats(acc); err != nil {
			return fmt.Errorf("gathering root stats failed: %v", err)
		}
	}
```

Reply:
I'm just trying to learn here. Would this also work?
```suggestion
	if h.gatherRoot; err := h.gatherRootStats(acc); err != nil {
		return fmt.Errorf("gathering root stats failed: %v", err)
	}
```

Reply:
Sure thing, changed.

Reply:
@Hipska No, this will not compile.